src/inspector_socket.cc

```diff
@@ -0,0 +1,762 @@
+    char extended_payload_length[8];
+    size_t remaining = data_length;
+    // Fill the length into extended_payload_length in the network byte order.
+    for (int i = 0; i < 8; ++i) {
+      extended_payload_length[7 - i] = remaining & 0xFF;
+      remaining >>= 8;
+    }
+    frame.insert(frame.end(), extended_payload_length,
+                 extended_payload_length + 8);
+    assert(!remaining);
+  }
+
+  if (masking_key != 0) {
+    const char* mask_bytes = reinterpret_cast<char*>(&masking_key);
+    frame.insert(frame.end(), mask_bytes, mask_bytes + 4);
+    for (size_t i = 0; i < data_length; ++i)  // Mask the payload.
+      frame.push_back(message[i] ^ mask_bytes[i % kMaskingKeyWidthInBytes]);
+  } else {
+    frame.insert(frame.end(), message, message + data_length);
+  }
+  *output = reinterpret_cast<char*>(malloc(frame.size() + 1));
+  memcpy(*output, &frame[0], frame.size());
+  (*output)[frame.size()] = '\0';
+  *output_len = frame.size();
+}
+
+static ws_decode_result decode_frame_hybi17(const char* buffer_begin,
+                                            size_t data_length,
+                                            bool client_frame,
+                                            int* bytes_consumed,
+                                            char** output,
+                                            bool* compressed) {
+  *bytes_consumed = 0;
+  *output = nullptr;
+  if (data_length < 2)
+    return FRAME_INCOMPLETE;
+
+  const char* p = buffer_begin;
+  const char* buffer_end = p + data_length;
+
+  unsigned char first_byte = *p++;
+  unsigned char second_byte = *p++;
+
+  bool final = (first_byte & kFinalBit) != 0;
+  bool reserved1 = (first_byte & kReserved1Bit) != 0;
+  bool reserved2 = (first_byte & kReserved2Bit) != 0;
+  bool reserved3 = (first_byte & kReserved3Bit) != 0;
+  int op_code = first_byte & kOpCodeMask;
+  bool masked = (second_byte & kMaskBit) != 0;
+  *compressed = reserved1;
+  if (!final || reserved2 || reserved3)
+    return FRAME_ERROR;  // Only compression extension is supported.
+
+  bool closed = false;
+  switch (op_code) {
+    case kOpCodeClose:
+      closed = true;
+      break;
+    case kOpCodeText:
+      break;
+    case kOpCodeBinary:        // We don't support binary frames yet.
+    case kOpCodeContinuation:  // We don't support binary frames yet.
+    case kOpCodePing:          // We don't support binary frames yet.
+    case kOpCodePong:          // We don't support binary frames yet.
+    default:
+      return FRAME_ERROR;
+  }
+
+  // In Hybi-17 spec client MUST mask its frame.
+  if (client_frame && !masked) {
+    return FRAME_ERROR;
+  }
+
+  uint64_t payload_length64 = second_byte & kPayloadLengthMask;
+  if (payload_length64 > kMaxSingleBytePayloadLength) {
+    int extended_payload_length_size;
+    if (payload_length64 == kTwoBytePayloadLengthField) {
+      extended_payload_length_size = 2;
+    } else {
+      assert(payload_length64 == kEightBytePayloadLengthField);
```

Comment:
ASSERT_EQ - although I don't think aborting on malformed external input is appropriate here.